File: 034_Cycles-n-Graphs.py
### Graphs: Cycle Detection (Python DSA)

---

### ✅ Cycle Detection Using DFS (Undirected Graph)
```python
class Graph:
    def __init__(self, size):
        self.adj_matrix = [[0] * size for _ in range(size)]
        self.size = size
        self.vertex_data = [''] * size

    def add_edge(self, u, v):
        self.adj_matrix[u][v] = 1
        self.adj_matrix[v][u] = 1  # undirected

    def add_vertex_data(self, vertex, data):
        self.vertex_data[vertex] = data

    def dfs_util(self, v, visited, parent):
        visited[v] = True
        for i in range(self.size):
            if self.adj_matrix[v][i] == 1:
                if not visited[i]:
                    if self.dfs_util(i, visited, v):
                        return True
                elif i != parent:
                    return True
        return False

    def is_cyclic(self):
        visited = [False] * self.size
        for i in range(self.size):
            if not visited[i]:
                if self.dfs_util(i, visited, -1):
                    return True
        return False

g = Graph(7)
for i, label in enumerate(['A', 'B', 'C', 'D', 'E', 'F', 'G']):
    g.add_vertex_data(i, label)

g.add_edge(3, 0)  # D - A
g.add_edge(0, 2)  # A - C
g.add_edge(0, 4)  # A - E
g.add_edge(4, 2)  # E - C
g.add_edge(2, 5)  # C - F
g.add_edge(2, 1)  # C - B
g.add_edge(2, 6)  # C - G
g.add_edge(1, 5)  # B - F

print("Graph has cycle:", g.is_cyclic())
```

---

### 🔁 Cycle Detection Using DFS (Directed Graph)
```python
class Graph:
    def __init__(self, size):
        self.adj_matrix = [[0] * size for _ in range(size)]
        self.size = size
        self.vertex_data = [''] * size

    def add_edge(self, u, v):
        self.adj_matrix[u][v] = 1  # directed only

    def add_vertex_data(self, vertex, data):
        self.vertex_data[vertex] = data

    def dfs_util(self, v, visited, recStack):
        visited[v] = True
        recStack[v] = True
        print("Current vertex:", self.vertex_data[v])

        for i in range(self.size):
            if self.adj_matrix[v][i] == 1:
                if not visited[i]:
                    if self.dfs_util(i, visited, recStack):
                        return True
                elif recStack[i]:
                    return True

        recStack[v] = False
        return False

    def is_cyclic(self):
        visited = [False] * self.size
        recStack = [False] * self.size
        for i in range(self.size):
            if not visited[i]:
                if self.dfs_util(i, visited, recStack):
                    return True
        return False

g = Graph(7)
for i, label in enumerate(['A', 'B', 'C', 'D', 'E', 'F', 'G']):
    g.add_vertex_data(i, label)

g.add_edge(3, 0)  # D -> A
g.add_edge(0, 2)  # A -> C
g.add_edge(2, 1)  # C -> B
g.add_edge(2, 4)  # C -> E
g.add_edge(1, 5)  # B -> F
g.add_edge(4, 0)  # E -> A
g.add_edge(2, 6)  # C -> G

print("Graph has cycle:", g.is_cyclic())
```

---

### 🔄 Cycle Detection Using Union-Find (Undirected Graph)
```python
class Graph:
    def __init__(self, size):
        self.adj_matrix = [[0] * size for _ in range(size)]
        self.size = size
        self.vertex_data = [''] * size
        self.parent = [i for i in range(size)]

    def add_edge(self, u, v):
        self.adj_matrix[u][v] = 1
        self.adj_matrix[v][u] = 1

    def add_vertex_data(self, vertex, data):
        self.vertex_data[vertex] = data

    def find(self, i):
        if self.parent[i] == i:
            return i
        return self.find(self.parent[i])

    def union(self, x, y):
        x_root = self.find(x)
        y_root = self.find(y)
        print(f'Union: {self.vertex_data[x]} + {self.vertex_data[y]}')
        self.parent[x_root] = y_root
        print(self.parent, '\n')

    def is_cyclic(self):
        for i in range(self.size):
            for j in range(i + 1, self.size):
                if self.adj_matrix[i][j]:
                    x = self.find(i)
                    y = self.find(j)
                    if x == y:
                        return True
                    self.union(x, y)
        return False

g = Graph(7)
for i, label in enumerate(['A', 'B', 'C', 'D', 'E', 'F', 'G']):
    g.add_vertex_data(i, label)

g.add_edge(1, 0)  # B - A
g.add_edge(0, 3)  # A - D
g.add_edge(0, 2)  # A - C
g.add_edge(2, 3)  # C - D
g.add_edge(3, 4)  # D - E
g.add_edge(3, 5)  # D - F
g.add_edge(3, 6)  # D - G
g.add_edge(4, 5)  # E - F

print("Graph has cycle:", g.is_cyclic())
```

---

File: 034_Cycles-n-Graphs.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Graph {
    int size;
    vector<vector<int>> adj_matrix;
    vector<string> vertex_data;

    // Helper DFS function for cycle detection
    bool dfs_util(int v, vector<bool>& visited, int parent) {
        visited[v] = true;
        for (int i = 0; i < size; i++) {
            if (adj_matrix[v][i] == 1) {
                if (!visited[i]) {
                    if (dfs_util(i, visited, v)) {
                        return true;
                    }
                }
                else if (i != parent) {
                    // Visited vertex that is not parent means cycle
                    return true;
                }
            }
        }
        return false;
    }

public:
    Graph(int sz) {
        size = sz;
        adj_matrix.resize(size, vector<int>(size, 0));
        vertex_data.resize(size, "");
    }

    void addEdge(int u, int v) {
        if (u >= 0 && u < size && v >= 0 && v < size) {
            adj_matrix[u][v] = 1;
            adj_matrix[v][u] = 1; // undirected graph
        }
    }

    void addVertexData(int vertex, const string& data) {
        if (vertex >= 0 && vertex < size) {
            vertex_data[vertex] = data;
        }
    }

    bool isCyclic() {
        vector<bool> visited(size, false);
        for (int i = 0; i < size; i++) {
            if (!visited[i]) {
                if (dfs_util(i, visited, -1)) {
                    return true;
                }
            }
        }
        return false;
    }

    void printGraph() {
        cout << "Adjacency Matrix:\n";
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                cout << adj_matrix[i][j] << " ";
            }
            cout << "\n";
        }
        cout << "\nVertex Data:\n";
        for (int i = 0; i < size; i++) {
            cout << i << ": " << vertex_data[i] << "\n";
        }
    }
};


int main() {
    Graph g(4);

    g.addVertexData(0, "A");
    g.addVertexData(1, "B");
    g.addVertexData(2, "C");
    g.addVertexData(3, "D");

    g.addEdge(0, 1);
    g.addEdge(0, 2);
    g.addEdge(1, 2);
    g.addEdge(2, 3);

    g.printGraph();

    if (g.isCyclic()) {
        cout << "Graph contains cycle\n";
    } else {
        cout << "Graph does not contain cycle\n";
    }

    return 0;
}
